testing/sync-test/src/auth.rs

```diff
@@ -180,29 +167,51 @@ pub struct TestClient {
 }
 
 impl TestClient {
-    pub fn new(acct: Arc<TestAccount>) -> Result<Self, failure::Error> {
+    pub fn new(acct: Arc<TestAccount>) -> Result<Self> {
         log::info!("Doing oauth flow!");
-
+        let auth_pw = auth::auth_pwd(&acct.email, &acct.pass);
         let mut fxa = FirefoxAccount::with_config(acct.cfg.clone());
         let oauth_uri = fxa.begin_oauth_flow(&[SYNC_SCOPE])?;
-        let auth_url = acct.cfg.auth_url()?;
-        let redirected_to = run_helper_command(
-            "oauth",
-            &[&acct.email, &acct.pass, auth_url.as_str(), &oauth_uri],
-        )?;
-
-        log::info!("Helper command gave '{}'", redirected_to);
+        let (session_token, key_fetch_token, verified) =
+            auth::account_login(&acct.cfg, &acct.email, &acct.pass).unwrap();
 
-        let final_url = Url::parse(&redirected_to)?;
-        let query_params = final_url
-            .query_pairs()
-            .into_owned()
-            .collect::<HashMap<String, String>>();
-
-        // should we be using the OAuthInfo this returns?
-        fxa.complete_oauth_flow(&query_params["code"], &query_params["state"])?;
+        let auth_key = http_client::derive_auth_key_from_session_token(&session_token)?;
+        if !verified {
+            log::warn!("Account needs verification... Verifying...");
+            Self::verify_account(&acct.email, &acct.cfg, &auth_key)?;
+        }
+        let (sync_key, xcs_key) =
+            auth::get_sync_keys(&acct.cfg, &key_fetch_token, &acct.email, &auth_pw)?;
+        let url = Url::parse(&oauth_uri)?;
+        let query_map: HashMap<String, String> = url.query_pairs().into_owned().collect();
+        let jwk_base_64 = query_map.get("keys_jwk").unwrap();
+        let decoded = base64::decode(&jwk_base_64).unwrap();
+        let jwk = std::str::from_utf8(&decoded)?;
+        let scope = query_map.get("scope").unwrap();
+        let client_id = query_map.get("client_id").unwrap();
+        let state = query_map.get("state").unwrap();
+        let code_challenge = query_map.get("code_challenge").unwrap();
+        let code_challenge_method = query_map.get("code_challenge_method").unwrap();
+        let keys_jwe = auth::create_keys_jwe(
+            &client_id,
+            &scope,
+            &jwk,
+            &auth_key,
+            &acct.cfg,
+            (&sync_key, &xcs_key),
+        )?;
+        let auth_params = auth::AuthorizationParameters::new(
+            client_id.clone(),
+            code_challenge.clone(),
+            code_challenge_method.clone(),
+            scope.clone(),
+            keys_jwe,
+            state.clone(),
+        );
+        let (code, state) = auth::send_authorization_request(&acct.cfg, auth_params, &auth_key)?;
```

Comment:
From a code-organization perspective, this key-fetching PKCE-code-authorizing stuff feels like it should mostly belong in the `TestAccount`, not the `TestClient`. Suggestion: make a `TestAccount::execute_oauth_flow` method that takes the URL returned by `fxa.begin_oauth_flow` here, does all the necessary authorization stuff, and returns the URI to which you would have been directed if this were a real OAuth flow. Encapsulating "all the bits that a user would normally do" inside the `TestAccount` seems like it will help us keep things straight in ours heads when we come to revisit this code in the future.